docs/graph-homomorphism-as-natural-transformation.md

```diff
@@ -66,6 +80,11 @@ F(t) -> snd
 ```
 
 (note: I believe that this definition requires that every vertex in the graph be both the source of an edge and destination of an edge)
+[Paul: This is related to a concern you expressed on slack: that the construction wouldn't work for DAG's or graphs with isolated vertices.  But everything is fine: the key point is that the functions F(s) and F(t) from the edge set to the vertex set need not be surjective.  I'll illustrate by means of an example.
+
+Consider the graph whose vertext set is V = {1,2,3,4} and whose edge set is E = {e=(1,2), f=(2,3)}.  So this is the path graph on the vertices 1,2,3 (a DAG) together with the isolated vertex 4.  Let us construct the functor X -> *Set* which corresponds to this graph.
+
+As usual we set F(a) = E and F(b) = V.  Define F(s): F(a) -> F(b) by F(s)(e) = 1 and F(s)(f) = 2, and define F(t): F(a) -> F(b) by F(t)(e) = 2 and F(t)(f) = 3. The vertices 3 and 4 do not lie in the range of F(s), and the vertices 1 and 4 do not lie in the range of F(t), but that's ok - these are still well-defined functions.]
```

Comment:
Ooh! This touches on a really interesting and important point: the words "injective" (AKA one-to-one), "surjective" (AKA onto), and "bijective (AKA one-to-one and onto) do *not* make sense in category theory!  This is because their definitions refer to the internal structure of an object (namely, the elements of a set), and as far as an abstract category is concerned the only structures that an object can possess are those which an be expressed externally using morphisms.  Instead, injective, surjective, and bijective really belong to set theory.  And there are some categories in which it is not possible to think of the objects as sets and the morphisms as functions between sets - the first example which comes to mind is the homotopy category, but maybe I'll try to come up with a more elementary example.

That said, there are *generalizations* of the words injective, surjective, and bijective to category theory: the generalizations are "monomorphism", "epimorphism", and "isomorphism", respectively.  A monomorphism is a morphism f: X -> Y with the following property: if g1, g2: Z -> X is a pair of morphisms such that the compositions f g1, f g2 are equal, then g1 = g2.  Similarly, f: X -> Y is an epimorphism if it has the following property: if g1, g2: Y -> Z is a pair of morphisms such taht the compositions g1 f, g2 f are equal, then g1 = g2.

Exercise: Show that in the category **Set** a morphism is mono if and only if it is an injective function, and a morphism is epi if and only if it is a surjective function.
Exercise: Show that an isomorphism is both mono and epi.
Exercise: Consider the poset category of natural numbers (see my post https://pwsiegel.github.io/category-theory/posets/ for definitions).  What are the monomorphisms?  What are the epimorphisms?  Are there morphisms which are mono and epi but not iso?

(In a certain kind of category called a *topos*, a morphism is mono and epi if and only if it is an iso.  Many important categories, including the category of graphs, are topoi.)

Finally, as you pointed out in your comment: functors do not necessarily preserve monomorphisms and epimorphisms!  You can see why this might happen from looking at the definitions: being mono or epi is all about having a cancellative property for other pairs of morphisms, so if you're in a category without many morphisms then it's easy to have that property.  But once a functor sends you into a category with lots of morphisms (like **Set**), it becomes a lot harder.  Indeed, in the category X (with two objects and two distinct morphisms between them), both morphisms s and t are mono and epi (but not iso) for trivial reasons, but as soon as they are mapped into **Set** they almost certainly are not injective or surjective.